_source/_posts/2021-09-12-spring-boot-spa.md

```diff
@@ -0,0 +1,686 @@
+
+Take a look at [this Spring blog post](https://spring.io/blog/2015/06/08/cors-support-in-spring-framework) for more info on Spring Boot and CORS.
+
+## Use The Okta CLI To Configure The Resource Server
+
+You should already have used the Okta CLI to either register a new account or log into an existing account. If not, please do so now. If you type `okta login`, you should see a message that says something like: `Okta Org already configured: https://dev-123456.okta.com/`.
+
+You are going to use the CLI to configure the Spring Boot project to use Okta for JWT authentication. The CLI will create an OpenID Connect (OIDC) application on Okta's servers that will be the authentication and authorization provider for the resource server.
+
+To start the configuration process, use the following command.
+
+```bash
+okta apps create
+```
+
+For `name`, type `spring-boot-spa-rest`.
+
+Enter `1` to specify a `Web` application.
+
+Enter `4` to specify `Other` as the type of web application.
+
+Press `enter` twice to accept the defaults for the redirect URIs.
+
+You can see the full console output below.
+
+```bash
+$: okta apps create
+Application name [spring-boot-spa]: spring-boot-spa-rest
+Type of Application
+(The Okta CLI only supports a subset of application types and properties):
+> 1: Web
+> 2: Single Page App
+> 3: Native App (mobile)
+> 4: Service (Machine-to-Machine)
+Enter your choice [Web]: 1
+Type of Application
+> 1: Okta Spring Boot Starter
+> 2: Spring Boot
+> 3: JHipster
+> 4: Other
+Enter your choice [Other]: 4
+Redirect URI
+Common defaults:
+ Spring Security - http://localhost:8080/login/oauth2/code/okta
+ JHipster - http://localhost:8080/login/oauth2/code/oidc
+Enter your Redirect URI [http://localhost:8080/callback]: 
+Enter your Post Logout Redirect URI [http://localhost:8080/]: 
+Configuring a new OIDC Application, almost done:
+Created OIDC application, client-id: 02b4cj4ys0G92cXYI4x7
+
+Okta application configuration has been written to: /home/andrewcarterhughes/Development/okta/2021/spring-boot-spa/spring-boot-spa/.okta.env
+```
+
+This creates a configuration file, `.okta.env` that you can source to configure the application.
+
+```bash
+source .okta.env
+```
+
+**You will need to source this file each time you open a new console** to run the application. 
+
+In a production scenario, you will need to ensure that these environment variables are loaded into the environment in which the Spring Boot application is running. **For security, it's also important that the file, and the client secret in particular, is not checked into source control or otherwise leaked.** Thus you may want to add `.okta.env` to the `.gitignore` of any similar projects you create (as I have here).
+
+The `.okta.env` file is simply a script that exports the necessary properties.
+
+```bash
+export OKTA_OAUTH2_ISSUER="https://dev-123456.okta.com/oauth2/default"
```

Comment:
You actually don't need to create an OAuth app on the Okta side when using JWT Access Tokens.
 Only the "issuer" is needed (you can stick it in the app.props file, or use the same environment variable).

That does create a chicken or an egg problem, as it's _nice_ to be able to copy the config values from the CLI,

One quick workaround someone can run `okta login` to display the current org:
> Okta Org already configured: https://dev-133337.okta.com


Longer explanation, For a resource server, the issuer allows for discovery of everything that is needed, except the expected values for the `aud` claim, which [defaults to `api://default`](https://github.com/okta/okta-spring-boot#configure-your-properties)